packages/doenetml-worker-rust/lib-doenetml-core/src/core/general_prop/component_ref_prop.rs

```diff
@@ -0,0 +1,95 @@
+use std::rc::Rc;
+
+use crate::{
+    components::prelude::*,
+    props::ContentFilter,
+    state::types::{component_refs::ComponentRef, content_refs::ContentRef},
+};
+
+/// A prop that references a single component
+#[derive(Debug)]
+pub struct ComponentRefProp {
+    /// The data query that indicates how the dependencies of this prop will be created.
+    data_query: DataQuery,
+
+    elements_to_select: Option<ElementsToSelect>,
+}
+
+#[derive(Debug)]
+enum ElementsToSelect {
+    First,
+    Last,
+}
+
+impl ComponentRefProp {
+    /// Creates a ComponentRefs prop that returns the last child with component_type
+    pub fn new_from_last_matching_child(component_type: &'static str) -> Self {
+        ComponentRefProp {
+            data_query: DataQuery::ComponentRefs {
+                container: PropSource::Me,
+                filter: Rc::new(ContentFilter::IsType(component_type)),
+            },
+            elements_to_select: Some(ElementsToSelect::Last),
+        }
+    }
+
+    /// Creates a ComponentRefs prop that returns the first child with component_type
+    pub fn new_from_first_matching_child(component_type: &'static str) -> Self {
+        ComponentRefProp {
+            data_query: DataQuery::ComponentRefs {
+                container: PropSource::Me,
+                filter: Rc::new(ContentFilter::IsType(component_type)),
+            },
+            elements_to_select: Some(ElementsToSelect::First),
+        }
+    }
+
+    pub fn new_self_ref() -> Self {
+        ComponentRefProp {
+            data_query: DataQuery::SelfRef,
+            elements_to_select: None,
+        }
+    }
+}
+
+impl PropUpdater for ComponentRefProp {
```

Comment:
It's interesting how here, and in `component_refs_prop`, we're not using the `RequiredData` structure. My inclination would be to convert these to use that structure and `TryFromDataQueryResults`, etc., macros as well, so that everything is consistent and we have a single pattern for developers to follow.

Reply:
I think the original reason it was different was because it was combined with data queries that might return different length vecs. I can switch to using `RequiredData` and see what happens.